crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,127 @@
+
+#[derive(Debug)]
+/// Struct to monitor network
+pub struct NetworkMonitor {
+    if_cache_updater_handle: Option<JoinHandle<io::Result<()>>>,
+    // Mac and Windows use callbacks, whereas in Linux a socket is registered
+    // with netlink multicast group messages
+    #[cfg(all(not(test), target_os = "linux"))]
+    monitor_handle: Option<JoinHandle<io::Result<()>>>,
+    #[cfg(all(not(test), target_os = "windows"))]
+    monitor_handle: crate::windows::SafeHandle,
+}
+
+impl NetworkMonitor {
+    /// Sets up and spawns network monitor
+    pub async fn new(if_addr: GetIfAddrs) -> std::io::Result<NetworkMonitor> {
+        *(LOCAL_ADDRS_CACHE.lock()) = local_interfaces::gather_local_interfaces(if_addr)?;
+
+        let if_cache_updater_handle = Some(tokio::spawn({
+            let mut notify = PATH_CHANGE_BROADCAST.subscribe();
+            async move {
+                loop {
+                    match notify.recv().await {
+                        Ok(()) => match local_interfaces::gather_local_interfaces(if_addr) {
+                            Ok(v) => {
+                                telio_log_debug!("Updating local addr cache");
+                                *(LOCAL_ADDRS_CACHE.lock()) = v;
+                            }
+                            Err(e) => telio_log_warn!("Unable to get local interfaces {e}"),
+                        },
+                        Err(e) => {
+                            telio_log_warn!("Failed to receive new path for sockets ({e}): ");
+                        }
+                    }
+                }
+            }
+        }));
+
+        // Only add in production, otherwise network triggers from OS can occur
+        // which can make the test (test_path_change_notification) flaky sometimes
+        #[cfg(all(
+            not(test),
+            any(target_os = "macos", target_os = "ios", target_os = "tvos")
+        ))]
+        NW_PATH_MONITOR_START.call_once(crate::apple::setup_network_monitor);
+        #[cfg(all(not(test), target_os = "linux"))]
+        let monitor_handle = crate::linux::setup_network_monitor().await;
+
+        #[cfg(all(not(test), target_os = "windows"))]
+        let monitor_handle = crate::windows::setup_network_monitor();
+
+        Ok(Self {
+            if_cache_updater_handle,
+            #[cfg(all(not(test), any(target_os = "linux", target_os = "windows")))]
+            monitor_handle,
+        })
+    }
+}
+
+impl Drop for NetworkMonitor {
+    fn drop(&mut self) {
+        if let Some(handle) = &self.if_cache_updater_handle {
+            handle.abort();
+        }
+        #[cfg(all(not(test), target_os = "linux"))]
+        if let Some(handle) = &self.monitor_handle {
+            handle.abort();
+        }
+        #[cfg(all(not(test), target_os = "windows"))]
+        if !self.monitor_handle.0.is_null() {
+            crate::windows::deregister_network_monitor(self.monitor_handle.clone());
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Please add more tests, for the init/deinit as we're providing custom constructor/destructor and specially for the `NetworkMonitor::if_cache_updated_handle` task:

```
loop {
    match notify.recv().await {
        Ok(()) => match local_interfaces::gather_local_interfaces(if_addr) {
            Ok(v) => {
                telio_log_debug!("Updating local addr cache");
                *(LOCAL_ADDRS_CACHE.lock()) = v;
            }
            Err(e) => telio_log_warn!("Unable to get local interfaces {e}"),
        },
        Err(e) => {
            telio_log_warn!("Failed to receive new path for sockets ({e}): ");
        }
    }
}
```